AI/VCAI/Pathfinding/AINodeStorage.h

```diff
@@ -0,0 +1,97 @@
+/*
+* AIhelper.h, part of VCMI engine
+*
+* Authors: listed in file AUTHORS in main folder
+*
+* License: GNU General Public License v2.0 or later
+* Full text of license available in license.txt file, in main folder
+*
+*/
+
+#pragma once
+
+#include "../../../lib/CPathfinder.h"
+#include "../../../lib/mapObjects/CGHeroInstance.h"
+#include "AIUtility.h"
+
+struct AIPathNode : public CGPathNode
+{
+	uint32_t chainMask;
+	uint64_t danger;
+};
+
+struct AIPathNodeInfo
+{
+	uint32_t movementPointsLeft;
+	uint32_t movementPointsUsed;
+	int turns;
+	int3 coord;
+	uint64_t danger;
+};
+
+struct AIPath
+{
+	std::vector<AIPathNodeInfo> nodes;
+
+	AIPath();
+
+	/// Gets danger of path excluding danger of visiting the target object like creature bank
+	uint64_t getPathDanger();
+
+	/// Gets danger of path including danger of visiting the target object like creature bank
+	uint64_t getTotalDanger(HeroPtr hero);
+
+	int3 firstTileToGet();
+
+	uint32_t movementCost();
+};
+
+class AINodeStorage : public INodeStorage
+{
+private:
+	mutable boost::mutex pathMx;
+	int3 sizes;
+	boost::multi_array<AIPathNode, 5> nodes; //[w][h][level][layer][chains]
```

Comment:
Here, that comment is also not too usefull, I even can't understand what means `[w][h]`?